packages/core/contracts/bot-helpers/uniswap-broker/UniswapBrokerV3.sol

```diff
@@ -0,0 +1,631 @@
+     */
+    function swapToPrice(
+        bool tradingAsEOA,
+        address uniswapPool,
+        address uniswapRouter,
+        uint160 sqrtRatioTargetX96,
+        address recipient,
+        uint256 deadline
+    ) external returns (uint256) {
+        // Create an instance of the pool and load in the current token price and the active tick.
+        IUniswapV3Pool pool = IUniswapV3Pool(uniswapPool);
+        (uint160 sqrtPriceX96, int24 tick, , , , , ) = pool.slot0();
+
+        // Work out the direction we need to trade. If the current price is more than the target price then we are
+        // trading token0 for token1. Else, we are trading token1 for token0.
+        bool zeroForOne = sqrtPriceX96 >= sqrtRatioTargetX96;
+
+        // Fetch the current pool liquidity. Build a state object to store this information which can be re-used during
+        // tick traversal later on.
+        uint128 startingLiquidity = pool.liquidity();
+        SwapState memory state =
+            SwapState({ sqrtPriceX96: sqrtPriceX96, tick: tick, liquidity: pool.liquidity(), requiredInputAmount: 0 });
+
+        // Iterate in a while loop that breaks when we hit the target price.
+        while (true) {
+            // Compute the next initialized tick. We only need to traverse initialized ticks as uninitialized ticks
+            // have the same liquidity as the previous tick.
+            StepComputations memory step;
+            step.sqrtPriceStartX96 = state.sqrtPriceX96;
+            (step.tickNext, step.initialized) = TickBitmap.nextInitializedTickWithinOneWord(
+                pool,
+                state.tick,
+                pool.tickSpacing(),
+                zeroForOne
+            );
+
+            // Double check we are not over or underflow the ticks.
+            if (step.tickNext < TickMath.MIN_TICK) step.tickNext = TickMath.MIN_TICK;
+            else if (step.tickNext > TickMath.MAX_TICK) step.tickNext = TickMath.MAX_TICK;
+
+            // Find the price at the next tick. Between the current state.sqrtPriceX96 and the nextTickPriceX96 we
+            // can find how much of the sold token is needed ot sufficiently move the market over the interval.
+
+            uint160 nextTickPriceX96 = TickMath.getSqrtRatioAtTick(step.tickNext);
+            uint256 inputAmountForStep;
+
+            // If zeroForOne is true, then we are moving the price UP. In this case we need to ensure that if the next
+            // tick price is more than the target price, we set the set the the next step price to the target price. This
+            // ensures that the price does not undershoot when the next tick is the last tick. Else, traverse the whole tick.
+            if (zeroForOne) {
+                step.sqrtPriceNextX96 = nextTickPriceX96 > sqrtRatioTargetX96 ? sqrtRatioTargetX96 : nextTickPriceX96;
+                inputAmountForStep = SqrtPriceMath.getAmount0Delta( // As we are trading token0 for token1, calculate the token0 input.
```

Comment:
Just curious, what is this function telling you? How much you need to put in to exhaust the tick? Or is it just telling you how much of token 0 is at the tick?

Reply:
it tells you the amount of token 0 you need to move the price from `step.sqrtPriceStartX96` to `step.sqrtPriceNextX96` for a given `state.liquidiy`. This method basically implements equation 5 from the PR description. it's not about exhausting a tick. rather, within a constant liquidity range, how much of token 0 you need to move the price.